File: Solutions/2265. Count Nodes Equal to Average of Subtree/2265. Count Nodes Equal to Average of Subtree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    array<int,3> dfs(TreeNode* root){
        if(!root){
            return {0,0,0};
        }
        auto p1=dfs(root->left),p2=dfs(root->right);
        int s=p1[0]+p2[0]+root->val,cnt=1+p1[1]+p2[1];
        return {s,cnt,p1[2]+p2[2]+(root->val==s/cnt)};
    }
    int averageOfSubtree(TreeNode* root) {
        return dfs(root)[2];
    }
};

File: Solutions/2265. Count Nodes Equal to Average of Subtree/2265. Count Nodes Equal to Average of Subtree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func dfs(root *TreeNode) []int {
    if root==nil{
        return []int{0,0,0}
    }
    p1,p2:=dfs(root.Left),dfs(root.Right)
    sum,cnt:=p1[0]+p2[0]+root.Val,1+p1[1]+p2[1]
    if(root.Val==sum/cnt){
        return []int{sum,cnt,p1[2]+p2[2]+1}
    }else{
        return []int{sum,cnt,p1[2]+p2[2]}
    }
}

func averageOfSubtree(root *TreeNode) int {
    return dfs(root)[2]
}